File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        int[] ans = {-1,-1};
        ArrayList<Integer> criticalPoints = new ArrayList<>();
        if(head == null || head.next == null || head.next.next ==  null){
            return ans;
        }
        int cnt = 1;
        ListNode prev = head;
        ListNode curr = head.next;
        while(curr.next != null){
            if(((prev.val<curr.val)&&(curr.val>curr.next.val)) || ((prev.val>curr.val)&&(curr.val<curr.next.val))){
                criticalPoints.add(cnt);
            }
            prev = curr;
            curr = curr.next;
            cnt++;
        }
        if(criticalPoints.size()<2){
            return ans;
        }

        int n = criticalPoints.size();
        int maxDiff = criticalPoints.get(n-1)-criticalPoints.get(0);
        int minDiff = Integer.MAX_VALUE;
        for(int i=1;i<n;i++){
            minDiff = Math.min(minDiff,criticalPoints.get(i)-criticalPoints.get(i-1));
        }
        ans[0] = minDiff;
        ans[1] = maxDiff;
        return ans;
    }
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        criticalPoints = []
        if head is None or head.next is None or head.next.next is None:
            return [-1,-1]
        temp = head.next
        prev = head
        cnt = 1
        while temp.next:
            if prev.val>temp.val<temp.next.val:
                criticalPoints.append(cnt)
            if prev.val<temp.val>temp.next.val:
                criticalPoints.append(cnt)
            prev = temp 
            temp=temp.next 
            cnt += 1
        if len(criticalPoints)<2:
            return [-1,-1]
        minDiff = float('inf')
        for i in range(1,len(criticalPoints)):
            minDiff = min(minDiff,criticalPoints[i]-criticalPoints[i-1])
        maxDiff = criticalPoints[-1]-criticalPoints[0]
        return [minDiff,maxDiff]

